### PR TITLE
Set of improvements of Sql translation subsystem

### PR DESCRIPTION
**Pull request by AlexUstinov**
This pull request removes unnecessary memory allocations as well as few cases of double processing.

### CODE REVIEW

File: Orm/Xtensive.Orm/Sql/Dml/SqlQueryRef.cs [hunk 1]
```diff
@@ -57,36 +56,36 @@ internal SqlQueryRef(ISqlQueryExpression query, string name)
...
+        if (queryExpression is SqlFreeTextTable freeTextTable ) {
+          queryColumns.AddRange(
+            freeTextTable.Columns.Select(originalColumn => SqlDml.TableColumn(this, originalColumn.Name)));
+        }
 
-        var containsTable = queryExpression as SqlContainsTable;
-        if (containsTable!=null )
-          foreach (var originalColumn in containsTable.Columns)
-            queryColumns.Add(SqlDml.TableColumn(this, originalColumn.Name));
+        if (queryExpression is SqlContainsTable containsTable ) {
```
[the remark this answers is not in the document]
Comment (AlexUstinov) [reply Orm/Xtensive.Orm/Sql/Dml/SqlQueryRef.cs]: Fixed

File: Orm/Xtensive.Orm/Sql/Dml/SqlQueryRef.cs [hunk 2]
```diff
@@ -57,36 +56,36 @@ internal SqlQueryRef(ISqlQueryExpression query, string name)
...
+            selectColumns[index] = column;
             queryColumns.Add(SqlDml.TableColumn(this, originalColumn.Name));
           }
         }
 
-        var freeTextTable = queryExpression as SqlFreeTextTable;
-        if (freeTextTable!=null )
-          foreach (var originalColumn in freeTextTable.Columns)
-            queryColumns.Add(SqlDml.TableColumn(this, originalColumn.Name));
+        if (queryExpression is SqlFreeTextTable freeTextTable ) {
```
[the remark this answers is not in the document]
Comment (AlexUstinov) [reply Orm/Xtensive.Orm/Sql/Dml/SqlQueryRef.cs]: Fixed

File: Orm/Xtensive.Orm/Sql/Dml/Collections/SqlTableColumnCollection.cs [hunk 3]
```diff
@@ -3,45 +3,99 @@
...
+        for (var index = count - 1; index >= 0; index--) {
+          var column = columnList[index];
+          var columnName = column.Name;
+          columnLookup[columnName] = column;
+          if (Comparer.Equals(columnName, name)) {
+            result = column;
+          }
+        }
+
+        return result;
```
[the remark this answers is not in the document]
Comment (AlexUstinov) [reply Orm/Xtensive.Orm/Sql/Dml/Collections/SqlTableColumnCollection.cs]: I've separated linear search and dictionary lookup into different methods.

File: Orm/Xtensive.Orm/Sql/Dml/SqlTableRef.cs [hunk 4]
```diff
@@ -17,75 +17,64 @@ namespace Xtensive.Sql.Dml
...
 
-      return clone;
-    }
+        clone.columns = new SqlTableColumnCollection(columnClones);
 
-    public override void AcceptVisitor(ISqlVisitor visitor)
-    {
-      visitor.Visit(this);
+        return clone;
+      }
```
[the remark this answers is not in the document]
Comment (AlexUstinov) [reply Orm/Xtensive.Orm/Sql/Dml/SqlTableRef.cs]: Thank you for pointing this out. This was an unfinished experiment. Assumption was additional block will not only restrict variable visibility but also will shift closure allocation point into this block. Assumption was wrong as it clear now, so I had to isolate clone creation part into separate method.

**Comment by alex-kulakov** [comment]
Thank you for corrections. PR is merged.

**Pull request merged**